In [1]:
import pandas as pd


In [2]:
datos = pd.read_csv('./dataset/credits.csv')

In [3]:
datos.head(1)

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   

                                                crew   id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...  862

In [4]:
print("Tipo de df:", type(datos))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


In [5]:
import numpy as np
import matplotlib.pylab as plt
import seaborn as sns 


In [6]:
datos.size

136428

In [7]:
dataset = pd.read_csv('./dataset/movies_dataset.csv')

C:\Users\Agustin\AppData\Local\Temp\ipykernel_11348\2553608563.py:1: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  dataset = pd.read_csv('./dataset/movies_dataset.csv')


#para poder ver columnas expandidas
# pd.options.display.max_columns=0

In [8]:
print(datos.shape)
print(dataset.shape)

(45476, 3)
(45466, 24)


In [9]:
datos.info

<bound method DataFrame.info of                                                     cast  \
0      [{'cast_id': 14, 'character': 'Woody (voice)',...   
1      [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2      [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3      [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4      [{'cast_id': 1, 'character': 'George Banks', '...   
...                                                  ...   
45471  [{'cast_id': 0, 'character': '', 'credit_id': ...   
45472  [{'cast_id': 1002, 'character': 'Sister Angela...   
45473  [{'cast_id': 6, 'character': 'Emily Shaw', 'cr...   
45474  [{'cast_id': 2, 'character': '', 'credit_id': ...   
45475                                                 []   

                                                    crew      id  
0      [{'credit_id': '52fe4284c3a36847f8024f49', 'de...     862  
1      [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...    8844  
2      [{'credit_id': '52fe466a9251416c75077a8

In [10]:
datos.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': "Savannah 'Vannah...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'credit_id': '52fe4284c3a36847f8024f49', 'de...    862  
1  [{'credit_id': '52fe44bfc3a36847f80a7cd1', 'de...   8844  
2  [{'credit_id': '52fe466a9251416c75077a89', 'de...  15602  
3  [{'credit_id': '52fe44779251416c91011acb', 'de...  31357  
4  [{'credit_id': '52fe44959251416c75039ed7', 'de...  11862

Para poder ver todas las columnas voy a hacer lo siguiente:

In [11]:
pd.options.display.max_columns = 0

In [12]:
dataset.head(1)

adult  ... vote_count
0  False  ...     5415.0

[1 rows x 24 columns]

Necesito entender què hay en el dataser:
 

In [13]:
dataset.columns

Index(['adult', 'belongs_to_collection', 'budget', 'genres', 'homepage', 'id',
       'imdb_id', 'original_language', 'original_title', 'overview',
       'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

In [14]:
dataset.dtypes

adult                     object
belongs_to_collection     object
budget                    object
genres                    object
homepage                  object
id                        object
imdb_id                   object
original_language         object
original_title            object
overview                  object
popularity                object
poster_path               object
production_companies      object
production_countries      object
release_date              object
revenue                  float64
runtime                  float64
spoken_languages          object
status                    object
tagline                   object
title                     object
video                     object
vote_average             float64
vote_count               float64
dtype: object

In [15]:
#belongs_to_collection está anidado
# tambièn (production_companies, genres, spoken_languages, production_countries) 

dataset['production_companies']

0           [{'name': 'Pixar Animation Studios', 'id': 3}]
1        [{'name': 'TriStar Pictures', 'id': 559}, {'na...
2        [{'name': 'Warner Bros.', 'id': 6194}, {'name'...
3        [{'name': 'Twentieth Century Fox Film Corporat...
4        [{'name': 'Sandollar Productions', 'id': 5842}...
                               ...                        
45461                                                   []
45462               [{'name': 'Sine Olivia', 'id': 19653}]
45463    [{'name': 'American World Pictures', 'id': 6165}]
45464                 [{'name': 'Yermoliev', 'id': 88753}]
45465                                                   []
Name: production_companies, Length: 45466, dtype: object

cada casa productora tiene un ID, pero leyendo lo que me están pidiendo que haga entiendo que no voy a usarlos. 

In [16]:
dataset['belongs_to_collection']

0        {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                      NaN
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                      NaN
4        {'id': 96871, 'name': 'Father of the Bride Col...
                               ...                        
45461                                                  NaN
45462                                                  NaN
45463                                                  NaN
45464                                                  NaN
45465                                                  NaN
Name: belongs_to_collection, Length: 45466, dtype: object

Eliminar las columnas que no serán utilizadas, video,imdb_id,adult,original_title,poster_path y homepage.

#antes de eliminar, voy a dejar marcadas las columnas que no voy
 a usar en el dataset original 
 
 video,imdb_id,adult,original_title,poster_path y homepage.

In [18]:
dataset[
    [
    #'adult', 
    'belongs_to_collection', 
    'budget', 
    'genres', 
    #'homepage', 
    'id',
    #'imdb_id',
    'original_language',
    #'original_title',
    'overview',
    'popularity',
    #'poster_path',
    'production_companies',
    'production_countries',
    'release_date',
    'revenue',
    'runtime',
    'spoken_languages',
    'status',
    'tagline',
    'title',
    #'video',
    'vote_average',
    'vote_count']
]

belongs_to_collection  ... vote_count
0      {'id': 10194, 'name': 'Toy Story Collection', ...  ...     5415.0
1                                                    NaN  ...     2413.0
2      {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...       92.0
3                                                    NaN  ...       34.0
4      {'id': 96871, 'name': 'Father of the Bride Col...  ...      173.0
...                                                  ...  ...        ...
45461                                                NaN  ...        1.0
45462                                                NaN  ...        3.0
45463                                                NaN  ...        6.0
45464                                                NaN  ...        0.0
45465                                                NaN  ...        0.0

[45466 rows x 18 columns]

podria usar 

dataset.drop(['', axis=1])  #axis 1 es para eliminar la columna, no la fila

pero es mejor copiar un nuevo dataset (o dframe) copiado como uno nuevo (usando .copy)

In [19]:
d_df = dataset[
    [
    #'adult', 
    'belongs_to_collection', 
    'budget', 
    'genres', 
    #'homepage', 
    'id',
    #'imdb_id',
    'original_language',
    #'original_title',
    'overview',
    'popularity',
    #'poster_path',
    'production_companies',
    'production_countries',
    'release_date',
    'revenue',
    'runtime',
    'spoken_languages',
    'status',
    'tagline',
    'title',
    #'video',
    'vote_average',
    'vote_count']
].copy()

chequeo que es de tipo dataframe

In [20]:
print("Tipo de df:", type(d_df))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


In [21]:
d_df.shape

(45466, 18)

En el enunciado me dice que los valores nulos de los campos revenue, budget deben ser rellenados por el número 0.

In [22]:
d_df['revenue'].isnull().sum()


6

In [23]:
d_df['budget'].isnull().sum()

0

confirmo que no cambié el tipo:

In [24]:
print("Tipo de df:", type(d_df))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


In [25]:
d_df.fillna({'revenue': 0}, inplace=True)

In [26]:
print("Tipo de df:", type(d_df))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


In [27]:
d_df['revenue'].isnull().sum()

0

Los valores nulos del campo release date deben eliminarse.

In [28]:
d_df['release_date'].shape

(45466,)

In [29]:
d_df['release_date'].isnull().sum()

87

In [30]:
df2= d_df.dropna(subset=['release_date'])

In [31]:
df2.info

<bound method DataFrame.info of                                    belongs_to_collection  ... vote_count
0      {'id': 10194, 'name': 'Toy Story Collection', ...  ...     5415.0
1                                                    NaN  ...     2413.0
2      {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...       92.0
3                                                    NaN  ...       34.0
4      {'id': 96871, 'name': 'Father of the Bride Col...  ...      173.0
...                                                  ...  ...        ...
45460                                                NaN  ...       26.0
45462                                                NaN  ...        3.0
45463                                                NaN  ...        6.0
45464                                                NaN  ...        0.0
45465                                                NaN  ...        0.0

[45379 rows x 18 columns]>

In [32]:
df2['release_date'].isnull().sum()

0

In [33]:
df2.shape

(45379, 18)

para ver si perdí datos:

In [34]:
d_df.shape

(45466, 18)

Comparando con fila 28 de codigo: no perdí datos.

df2 es el data frame con los arreglos:

Los valores nulos de los campos revenue, budget deben ser rellenados por el número 0.(prueba)

Los valores nulos del campo release date deben eliminarse. (df2)

Eliminar las columnas que no serán utilizadas, video,imdb_id,adult,original_title,poster_path y homepage. (d_df)



De haber fechas, deberán tener el formato AAAA-mm-dd, además deberán crear la columna release_year donde extraerán el año de la fecha de estreno.

In [35]:
print(type(df2))

<class 'pandas.core.frame.DataFrame'>


In [36]:
df2['release_date']

0        1995-10-30
1        1995-12-15
2        1995-12-22
3        1995-12-22
4        1995-02-10
            ...    
45460    1991-05-13
45462    2011-11-17
45463    2003-08-01
45464    1917-10-21
45465    2017-06-09
Name: release_date, Length: 45379, dtype: object

La columna tiempo es de tipo objeto

In [37]:
df_dt =df2.copy()

In [38]:
from datetime import datetime

In [39]:
#df_dt['release_date'] = pd.to_datetime(df_dt['release_date'])

para manejar este error, quese debe a que uno o más valores en tu columna no coinciden con el formato especificado, uso  errors='coerce' en el código.

In [40]:
df_dt['release_date'] = pd.to_datetime(df_dt['release_date'], format='%Y-%m-%d', errors='coerce')

In [41]:
df_dt['release_date']

0       1995-10-30
1       1995-12-15
2       1995-12-22
3       1995-12-22
4       1995-02-10
           ...    
45460   1991-05-13
45462   2011-11-17
45463   2003-08-01
45464   1917-10-21
45465   2017-06-09
Name: release_date, Length: 45379, dtype: datetime64[ns]

ahora es de datatime y tieme el formato AAAA-MM-DD.
Recien ahora puedo extraer una columna con el año, "release_year"

In [42]:
df_dt['release_year']=df_dt['release_date'].dt.year

In [43]:
df_dt['release_year']

0        1995.0
1        1995.0
2        1995.0
3        1995.0
4        1995.0
          ...  
45460    1991.0
45462    2011.0
45463    2003.0
45464    1917.0
45465    2017.0
Name: release_year, Length: 45379, dtype: float64

In [44]:
df_dt.head(1)

belongs_to_collection  ... release_year
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...       1995.0

[1 rows x 19 columns]

In [45]:
#df_dt['return'] = np.where(df_dt['revenue'].isna() | df_dt['budget'].isna() | (df_dt['budget'] == 0), 0, df_dt['revenue'] / df_dt['budget'])

tengo que hacer que todas tengan el formato float o int

In [46]:
print(df_dt.dtypes)

belongs_to_collection            object
budget                           object
genres                           object
id                               object
original_language                object
overview                         object
popularity                       object
production_companies             object
production_countries             object
release_date             datetime64[ns]
revenue                         float64
runtime                         float64
spoken_languages                 object
status                           object
tagline                          object
title                            object
vote_average                    float64
vote_count                      float64
release_year                    float64
dtype: object


budget es un objeto, y revenue es float.-

In [47]:
#df_dt['budget'] = df_dt['budget'].astype(float)

veo que hay valores que no me permiten seguir, los copio del mensaje de error y los saco:

 '/ff9qCepilowshEtG2GYWwzt2bs4.jpg' ese valor està en la columna

In [48]:
df_dt.loc[df_dt['budget'] == '/ff9qCepilowshEtG2GYWwzt2bs4.jpg', 'budget'] = np.nan

ahora debo eliminar el nan ese que cree para eliminar el jpg.

In [49]:
df_dt['budget'].isnull().sum()

1

In [50]:
df_dt.fillna({'budget': 0}, inplace=True)

vuelvo a intentar

In [51]:
#df_dt['budget'] = df_dt['budget'].astype(float)

aun falta este '/zV8bHuSL6WXoD6FWogP9j4x80bL.jpg'

In [52]:
df_dt.loc[df_dt['budget'] == '/zV8bHuSL6WXoD6FWogP9j4x80bL.jpg', 'budget'] = np.nan

In [53]:
#df_dt['budget'] = df_dt['budget'].astype(float)

falta este '/zaSf5OG7V8X8gqFvly88zDdRm46.jpg'

In [54]:
df_dt.loc[df_dt['budget'] == '/zaSf5OG7V8X8gqFvly88zDdRm46.jpg', 'budget'] = np.nan

In [55]:
df_dt['budget'] = df_dt['budget'].astype(float)
df_dt['revenue'] = df_dt['revenue'].astype(float)

In [56]:
df_dt.fillna({'budget': 0}, inplace=True)

veo si siguen budget como objeto y revenue como float.-

In [57]:
df_dt.dtypes

belongs_to_collection            object
budget                          float64
genres                           object
id                               object
original_language                object
overview                         object
popularity                       object
production_companies             object
production_countries             object
release_date             datetime64[ns]
revenue                         float64
runtime                         float64
spoken_languages                 object
status                           object
tagline                          object
title                            object
vote_average                    float64
vote_count                      float64
release_year                    float64
dtype: object

In [58]:
df_dt['return'] = np.where(df_dt['revenue'].isna() | df_dt['budget'].isna() | (df_dt['budget'] == 0), 0, df_dt['revenue'] / df_dt['budget'])

In [59]:
df_dt.head(1)

belongs_to_collection  ...     return
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  12.451801

[1 rows x 20 columns]

necesito dejar los valores como float, porque es una razon (una proporcion de retorno). Si pongo en INT el resultado me da un valor 12 y sin decimales.

para desanidar puedo usar explode() para listas, pero aqui tengo un diccionario. 

Primero voy a crear un df_ds para trabajar ese paso ahi.

df_info = pd.json_normalize(df['info'])
# Combinar el DataFrame original con el DataFrame desanidado
df_combined = df.drop('info', axis=1).join(df_info)

In [60]:
df_ds = df_dt.copy()

In [61]:
df_ds['belongs_to_collection']

0        {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                      NaN
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                      NaN
4        {'id': 96871, 'name': 'Father of the Bride Col...
                               ...                        
45460                                                  NaN
45462                                                  NaN
45463                                                  NaN
45464                                                  NaN
45465                                                  NaN
Name: belongs_to_collection, Length: 45379, dtype: object

In [62]:
df_ds['belongs_to_collection'].isnull().sum()

40888

In [63]:
df_ds= df_ds.dropna(subset=['belongs_to_collection'])

In [64]:
df_ds['belongs_to_collection'].isnull().sum()

0

In [65]:
df_ds['belongs_to_collection']

0        {'id': 10194, 'name': 'Toy Story Collection', ...
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
4        {'id': 96871, 'name': 'Father of the Bride Col...
9        {'id': 645, 'name': 'James Bond Collection', '...
12       {'id': 117693, 'name': 'Balto Collection', 'po...
                               ...                        
45355    {'id': 37261, 'name': 'The Carry On Collection...
45358    {'id': 37261, 'name': 'The Carry On Collection...
45369    {'id': 37261, 'name': 'The Carry On Collection...
45371    {'id': 477208, 'name': 'DC Super Hero Girls Co...
45382    {'id': 200641, 'name': 'Red Lotus Collection',...
Name: belongs_to_collection, Length: 4491, dtype: object

quedaron 4491 filas, antes habia 45379. Eso puede traer problemas creo


In [66]:
df_ds2 = df_dt.copy()

In [67]:
def replace_nan(value):
    if isinstance(value, dict):
        return {k: (0 if pd.isna(v) else v) for k, v in value.items()}
    elif isinstance(value, list):
        return [0 if pd.isna(i) else i for i in value]
    else:
        return value

df_ds2['belongs_to_collection'] = df_ds2['belongs_to_collection'].apply(replace_nan)

In [68]:
df_ds2.head(10)

belongs_to_collection  ...     return
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  12.451801
1                                                NaN  ...   4.043035
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...  ...   0.000000
3                                                NaN  ...   5.090760
4  {'id': 96871, 'name': 'Father of the Bride Col...  ...   0.000000
5                                                NaN  ...   3.123947
6                                                NaN  ...   0.000000
7                                                NaN  ...   0.000000
8                                                NaN  ...   1.838576
9  {'id': 645, 'name': 'James Bond Collection', '...  ...   6.072311

[10 rows x 20 columns]

In [69]:
df_ds['belongs_to_collection']

0        {'id': 10194, 'name': 'Toy Story Collection', ...
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
4        {'id': 96871, 'name': 'Father of the Bride Col...
9        {'id': 645, 'name': 'James Bond Collection', '...
12       {'id': 117693, 'name': 'Balto Collection', 'po...
                               ...                        
45355    {'id': 37261, 'name': 'The Carry On Collection...
45358    {'id': 37261, 'name': 'The Carry On Collection...
45369    {'id': 37261, 'name': 'The Carry On Collection...
45371    {'id': 477208, 'name': 'DC Super Hero Girls Co...
45382    {'id': 200641, 'name': 'Red Lotus Collection',...
Name: belongs_to_collection, Length: 4491, dtype: object

Solo el nmbre de la coleccion puede servir, creo

tambien quedaron 4491


In [70]:
from pandas import Series,DataFrame 

In [71]:
df_df_belongs = pd.DataFrame(df_ds2['belongs_to_collection'])

In [72]:
df_df_belongs.head()

belongs_to_collection
0  {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                NaN
2  {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                NaN
4  {'id': 96871, 'name': 'Father of the Bride Col...

In [73]:
print("Tipo de df:", type(df_df_belongs))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


convierto a dataframe para poder agrupar

In [74]:
print(df_df_belongs)

                                   belongs_to_collection
0      {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                    NaN
2      {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                    NaN
4      {'id': 96871, 'name': 'Father of the Bride Col...
...                                                  ...
45460                                                NaN
45462                                                NaN
45463                                                NaN
45464                                                NaN
45465                                                NaN

[45379 rows x 1 columns]


In [75]:
#parece un jason adentro de la columna
import json

puedo separar esas columnas anidadas en json y trabajarlas

In [76]:
#df_df_belongs.df_json('./belongs.json')

In [77]:
# leo JSON 
file = 'belongs.json'
with open(file) as train_file:
    dict_train = json.load(train_file)

# converting json dataset from dictionary to dataframe
train = pd.DataFrame.from_dict(dict_train, orient='index')
train.reset_index(level=0, inplace=True)

In [78]:
train.head()

index  ... 45465
0  belongs_to_collection  ...  None

[1 rows x 45380 columns]

Como no funciona, voy a intentar con explode()

Al igual que con los métodos stack(), unstack() y melt() vistos anteriormente,


necesito hacer que sea un formato de lista de python

In [80]:
df_collection = df_ds2["belongs_to_collection"].explode()

In [81]:
df_collection


0        {'id': 10194, 'name': 'Toy Story Collection', ...
1                                                      NaN
2        {'id': 119050, 'name': 'Grumpy Old Men Collect...
3                                                      NaN
4        {'id': 96871, 'name': 'Father of the Bride Col...
                               ...                        
45460                                                  NaN
45462                                                  NaN
45463                                                  NaN
45464                                                  NaN
45465                                                  NaN
Name: belongs_to_collection, Length: 45379, dtype: object

¿còmo hago para diferenciar el contenido de name? de un mensaje de este tipo:

'id': 10194, 
'name': 'Toy Story Collection', 
'poster_path': '\/7G9915LfUQ2lVfwMEEhDsn3kT4B.jpg', 
'backdrop_path': '\/9FBwqcd9IRruEDUrTdcaafOMKUq.jpg'


In [2]:
df_pelis = df_collection.copy()

NameError: name 'df_collection' is not defined

In [1]:
#estoy sin poder armar el código, voy a usar ChatGPT

def desanidarColumna(df, columna, key=None):
    def desanidarElemento(elemento, key):
        if isinstance(elemento, str):
            try:
                elemento = json.loads(elemento)
            except (ValueError, json.JSONDecodeError):
                return None
        if isinstance(elemento, list):
            if key:
                return ', '.join(str(d.get(key, '')) for d in elemento if isinstance(d, dict))
            return ', '.join(str(d) for d in elemento)
        if isinstance(elemento, dict):
            if key:
                return elemento.get(key, None)
            return str(elemento)
        return None

    return df[columna].apply(lambda x: desanidarElemento(x, key))

df_pelis['belongs_to_collection'] = desanidarColumna(df_pelis, 'belongs_to_collection')

Estoy tardando mucho, voy a trabajar en power BI y regresar lo desanidado aqui

In [82]:
df_collection.to_csv("collection.csv")

In [83]:
collection_name = pd.read_csv('collection_filtrado.csv')

In [84]:
collection_name

index          belongs_to_collection.3.1
0          0             'Toy Story Collection'
1          1           'No assigned Collection'
2          2        'Grumpy Old Men Collection'
3          3           'No assigned Collection'
4          4   'Father of the Bride Collection'
...      ...                                ...
29995  30026           'No assigned Collection'
29996  30027           'No assigned Collection'
29997  30028           'No assigned Collection'
29998  30029           'No assigned Collection'
29999  30030           'No assigned Collection'

[30000 rows x 2 columns]

A las columnas que estaban vacías le puse "Not assigned"

Como queda muy grande, el proceso en power BI le sacó 15 mil filas al exportar. Para mi trabajo MVP, de todas formas voy a tener que reducir el dataset para que funcione en FastAPI, por lo que usarè el criterio de tener màs de 20 filas, pero menos de 30, en base a la condicion de uno de los endpoints (tenès màs de 20000 votos para poder ser evaluado)

ahora tengo que hacer lo mismo para las otras columnas anidadas.

In [85]:
df_ds2['genres'].isnull().sum()


0

In [86]:
df_ds_genres = pd.DataFrame(df_ds2['genres'])

df_ds_genres.to_csv("genres2.csv")

trabajo en power BI

Elijo separar los gèneros, porque resultan valores muy importantes para recomendar películas si se los considera por separado.
Drama, Crimen, Familia, etc. resulta de gran valor por separado, pero juntos creo que no, ya que la combinación quizá no se da casi nunca.
Dejé solo tres gèneros (tres columnas con géneros), porque a partir del cuarto tengo mas del 80% de valores nulos.

In [87]:
genres_names = pd.read_csv('genres_filtrado.csv')

In [88]:
genres_names.head()

index      genres 1      genres 2      genres 3
0      0   'Animation'      'Comedy'      'Family'
1      1   'Adventure'     'Fantasy'      'Family'
2      2     'Romance'      'Comedy'  Not assigned
3      3      'Comedy'       'Drama'     'Romance'
4      4      'Comedy'  Not assigned  Not assigned

In [3]:
#estoy pensando si el "not assigned" puede ser considerado 
#como una coincidencia que va a afectar a mi recomendacion
#recordar que cuando sea not assigned, debo ignorar

teníamos:

In [90]:
df_ds2.head(1)

belongs_to_collection  ...     return
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  12.451801

[1 rows x 20 columns]

faltan desanidar cosas:

In [91]:
df_ds_languages = pd.DataFrame(df_ds2['spoken_languages'])
df_ds_languages.to_csv("languages.csv")


In [92]:
language_name = pd.read_csv('main_language.csv')

In [93]:
language_name

index Main_spoken_languages
0          0             'English'
1          1             'English'
2          2             'English'
3          3             'English'
4          4             'English'
...      ...                   ...
29995  30026             'Pусский'
29996  30027             'Pусский'
29997  30028             'Pусский'
29998  30029             'Pусский'
29999  30030             'ქართული'

[30000 rows x 2 columns]

NO voy a usar las productoras ni el idioma como criterio para recomendar

In [94]:
#spoken_languages, production_countries y production_companies

['spoken_languages', 'production_countries', 'production_companies']

In [95]:
df_ds2.columns

Index(['belongs_to_collection', 'budget', 'genres', 'id', 'original_language',
       'overview', 'popularity', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'vote_average',
       'vote_count', 'release_year', 'return'],
      dtype='object')

In [4]:
#voy a dejar afuera las cosas que no serán útiles para estas funciones que me piden

df_ds_dropeados = df_ds2[
    [
    'belongs_to_collection', 
    'budget',
    'genres', 
    'id', 
    #'original_language',
    'overview',
    'popularity',
    #'production_companies',
    #'production_countries',
    'release_date',
    'revenue',
    'runtime',
    'spoken_languages',
    'status',
    'tagline',
    'title',
    'vote_average',
    'vote_count',
    'release_year',
    'return'
    ]
].copy()

NameError: name 'df_ds2' is not defined

In [97]:
df_ds_dropeados.head(1)

belongs_to_collection  ...     return
0  {'id': 10194, 'name': 'Toy Story Collection', ...  ...  12.451801

[1 rows x 17 columns]

ahora reemplazo el belongs_to_collection, el genres y spoken languages

sacar spoken_languages
sacar belongs_to_collection
sacar genres


In [98]:
df_ds_dropeados.columns

Index(['belongs_to_collection', 'budget', 'genres', 'id', 'overview',
       'popularity', 'release_date', 'revenue', 'runtime', 'spoken_languages',
       'status', 'tagline', 'title', 'vote_average', 'vote_count',
       'release_year', 'return'],
      dtype='object')

In [99]:
df_ds3 = df_ds_dropeados[
[
 #'belongs_to_collection', 
 'budget', 
 #'genres', 
 'id', 
 'overview',
 'popularity', 
 'release_date',
 'revenue',
 'runtime',
 #'spoken_languages',
 'status',
 'tagline',
 'title',
 'vote_average',
 'vote_count',
 'release_year',
 'return']
].copy()

In [100]:
#a df_ds3 le voy a poner las columnas trabajadas
# language_name , collection_name , genres_names 
#o todo junto que es completos_desanidados.csv

In [101]:
nuevo_df = pd.read_csv('Completo_desanidados.csv')

In [102]:
nuevo_df.head(1)

index Main_spoken_languages  ...   genres 2   genres 3
0      0             'English'  ...   'Comedy'   'Family'

[1 rows x 6 columns]

In [103]:
print("Tipo de df:", type(nuevo_df))

Tipo de df: <class 'pandas.core.frame.DataFrame'>


In [104]:
df_final = df_ds3.join(nuevo_df)

Finalmente tengo algo así:

In [105]:
df_final.head(5)

budget     id  ...      genres 2      genres 3
0  30000000.0    862  ...      'Comedy'      'Family'
1  65000000.0   8844  ...     'Fantasy'      'Family'
2         0.0  15602  ...      'Comedy'  Not assigned
3  16000000.0  31357  ...       'Drama'     'Romance'
4         0.0  11862  ...  Not assigned  Not assigned

[5 rows x 20 columns]

In [106]:
#me falta ver si hace falta agregar la otra tabla

In [107]:
credits.columns

Index(['cast', 'crew', 'id'], dtype='object')

In [108]:
nuevo_credits_df = pd.read_csv('./dataset/credits_filtrado.csv')


Desanido datos en power Query, elijo quedarme con los 4 actores principales solamente (no todo el cast) y el director/a
, para poder cumplir con el requisito del MVP

In [109]:
nuevo_credits_df.head()

Cast       Directed by  ID
0      Tom Hanks, Tim Allen, Don Rickles, Jim Varney     John Lasseter   0
1  Robin Williams, Jonathan Hyde, Kirsten Dunst, ...      Joe Johnston   1
2  Walter Matthau, Jack Lemmon, Ann-Margret, Soph...     Howard Deutch   2
3  Whitney Houston, Angela Bassett, Loretta Devin...   Forest Whitaker   3
4  Steve Martin, Diane Keaton, Martin Short, Kimb...     Charles Shyer   4

Voy a probar una de lsaa funciones aquì, antes de pasar fastAPI, para asegurarme que mi dataset cargado responde a la necesidad:

In [110]:
df_release_month = pd.DatetimeIndex(df_final["release_date"]).month

In [111]:
m=5
print((pd.DatetimeIndex(df_final["release_date"]).month == m).sum())

3339


In [112]:
m=5
print(f"El número de estrenos realizados en el mes {m} es:",(pd.DatetimeIndex(df_final["release_date"]).month == m).sum())

El número de estrenos realizados en el mes 5 es: 3339


junto el df de cast con el df_final y lo paso a parquet para que sea más liviano. (para probar voy a usar un fragmento del dataset solamente)

In [113]:
#nuevo_credits_df (ID) + df_final (index)

df_completo = df_final.join(nuevo_credits_df)

In [114]:
df_completo.drop_duplicates()
df_completo.dropna()


budget      id  ...         Directed by       ID
1      65000000.0    8844  ...        Joe Johnston      1.0
2             0.0   15602  ...       Howard Deutch      2.0
3      16000000.0   31357  ...     Forest Whitaker      3.0
4             0.0   11862  ...       Charles Shyer      4.0
5      60000000.0     949  ...        Michael Mann      5.0
...           ...     ...  ...                 ...      ...
29993         0.0   73369  ...   Alessandro Capone  29993.0
29995         0.0   52051  ...         Alex Winter  29995.0
29996         0.0   78552  ...      Kabir Sadanand  29996.0
29998         0.0  320703  ...          Dana Brown  29998.0
29999         0.0  170277  ...          John McKay  29999.0

[14908 rows x 23 columns]

tengo casi 30 mil filas completas. Desde ahí había muchos valores nulos, a veces mas de la mitad, por eso  y por las razones que ya enuncie (debo hacer correr en fastAPIy despues en Render un MVP) voy a reducir a un set de 2000 para trabajar las pruebas.

In [115]:
df_reducido=df_completo.truncate(before= 1001, after=3001)

In [116]:
df_reducido.columns

Index(['budget', 'id', 'overview', 'popularity', 'release_date', 'revenue',
       'runtime', 'status', 'tagline', 'title', 'vote_average', 'vote_count',
       'release_year', 'return', 'index', 'Main_spoken_languages',
       'belongs_to_collection.3.1', 'genres 1', 'genres 2', 'genres 3', 'Cast',
       'Directed by', 'ID'],
      dtype='object')

In [117]:
df_reducido.head()

budget     id  ...        Directed by      ID
1001  10000000.0  11114  ...        Don Chaffey  1001.0
1002  20000000.0  12335  ...   Robert Stevenson  1002.0
1003   3000000.0  12092  ...     Hamilton Luske  1003.0
1004  12000000.0  10948  ...         Ted Berman  1004.0
1005   3000000.0  11229  ...     Matthew Bright  1005.0

[5 rows x 23 columns]

In [118]:
df_reducido_prueba = df_reducido.to_parquet("./dataset/df_reducido_prueba.parquet")

In [121]:
df_export = df_completo.to_csv("./dataset/df_completo.csv")

paso a redactar el main.py para fastAPI